### PR TITLE
Add blockVisibility for parentActivity blocks during creation and del…

### PR DESCRIPTION
**Pull request by YuriyPavlovBI**
Added reevaluation of the block visibilities for the parent activity when child activity is being added/removed. (use case: we need to hide "I don't have children" checkbox if at least one child is added)

### CODE REVIEW

File: pepper-apis/docs/specification/src/components/responses.yml [hunk 1]
```diff
@@ -35,18 +35,7 @@ ActivityAnswersPatchResponse:
...
-                - blockGuid
-                - shown
-              properties:
-                blockGuid:
-                  type: string
-                  description: identifier for the block
-                shown:
-                  type: boolean
-                  description: whether block is shown or not
+            $ref: '../pepper.yml#/components/schemas/Block.Visibility'
```
Comment (bskinner) [inline pepper-apis/docs/specification/src/components/responses.yml]: This is an improperly defined array type. An array's items may take a `$ref`, but it should be in an `items` map.

Either review the [array](https://swagger.io/docs/specification/data-models/data-types/#array) data type, or look at one of the other examples of an array definition in this specification.

File: pepper-apis/docs/specification/src/components/responses.yml [hunk 2]
```diff
@@ -111,6 +100,19 @@ ActivitySummaryResponse:
...
+    application/json:
+      schema:
+        type: object
+        required:
+          - blockVisibility
+        properties:
+          blockVisibility:
+            type: array
+            description: list of updated parent instance block visibilities
+            $ref: '../pepper.yml#/components/schemas/Block.Visibility'
```
Comment (bskinner) [inline pepper-apis/docs/specification/src/components/responses.yml]: This is an improperly defined array.

File: pepper-apis/docs/specification/src/components/schemas.yml [hunk 3]
```diff
@@ -755,6 +755,19 @@ Block.Content:
           type: string
         body:
           type: string
+Block.Visibility:
+  items:
```
Comment (bskinner) [inline pepper-apis/docs/specification/src/components/schemas.yml]: This is an improperly defined [Schema](https://swagger.io/specification/#schema-object) object. The `items` key belongs in the array definition, not here.

File: pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml [hunk 4]
```diff
@@ -61,17 +61,22 @@ post:
...
                 - instanceGuid
+                - blockVisibility
               properties:
                 instanceGuid:
                   type: string
                   example: ABCDEFGH12
+                blockVisibility:
+                  type: array
+                  description: list of updated parent instance block visibilities
+                  $ref: '../pepper.yml#/components/schemas/Block.Visibility'
```
Comment (bskinner) [inline pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml]: This is an improperly defined array.

**Review by bskinner — DISMISSED** [review]
Still doing a deeper review of the code, but I identified some issues in the OpenAPI spec changes.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java [hunk 5]
```diff
@@ -92,4 +105,36 @@ public static Answer getAnswer(Handle handle, long instanceId, String questionSt
                 .orElseThrow(() -> new DaoException(format(
                         "Error to detect answer: question stableId=%s, instanceId=%d", questionStableId, instanceId)));
     }
+
+    public static Optional<List<BlockVisibility>> getBlockVisibility(Handle handle, Response response,
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: For what I am seeing, in all the API changes the array BlockVisibilities is always required/not nullable property. Returning optional seems like it complicates the caller codes with conditionals. Might as well simply return an list every time, even if it is just empty.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java [hunk 6]
```diff
@@ -92,4 +105,36 @@ public static Answer getAnswer(Handle handle, long instanceId, String questionSt
...
+                                                                     String parentInstanceGuid, User participantUser, StudyDto studyDto,
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: Would it make sense to rename this argument to activityInstanceGuid? Just that, in the context of family history we do have activity instances that contain activity instances, so the `parent` prefix makes me think about that and don't think that is the intent at all.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java [hunk 7]
```diff
@@ -92,4 +105,36 @@ public static Answer getAnswer(Handle handle, long instanceId, String questionSt
...
+        if (parentInstanceDto != null) {
+
+            ActivityDefStore activityStore = ActivityDefStore.getInstance();
+            FormActivityDef parentActivity = ActivityInstanceUtil
+                    .getActivityDef(handle, activityStore, parentInstanceDto, studyDto.getGuid());
+
+            UserActivityInstanceSummary instanceSummary = RouteUtil.findUserActivityInstanceSummaryOrHalt(
+                    response, handle, participantUser.getGuid(), studyDto.getGuid(), parentInstanceGuid, isStudyAdmin);
+            result = formService.getBlockVisibilities(handle, instanceSummary, parentActivity,
+                    participantUser.getGuid(), operatorGuid, parentInstanceGuid);
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: Please verify, but it looks like the 3rd argument to `getBlockVisibilities` method ultimately is not used for anything in FormActivityService. If this is true, please refactor the argument out of the affected methods.

Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: Fixed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java (hunk 6 above)
Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: Fixed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java (hunk 5 above)
Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: Fixed

File: pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml (hunk 4 above)
Comment (YuriyPavlovBI) [reply pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml]: Fixed

File: pepper-apis/docs/specification/src/components/schemas.yml (hunk 3 above)
Comment (YuriyPavlovBI) [reply pepper-apis/docs/specification/src/components/schemas.yml]: Fixed

File: pepper-apis/docs/specification/src/components/responses.yml (hunk 2 above)
Comment (YuriyPavlovBI) [reply pepper-apis/docs/specification/src/components/responses.yml]: Fixed

File: pepper-apis/docs/specification/src/components/responses.yml (hunk 1 above)
Comment (YuriyPavlovBI) [reply pepper-apis/docs/specification/src/components/responses.yml]: Fixed

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**